File: example/project-nextjs-tailwind/components/atoms/input.js
// Atom/Input

// ### Variables
// name              | type        | default        | options
// ------------------------------------------------------------------
// arrow               boolean       true
// disabled            boolean       false
// errorMessage        string        ""
// hintText            string        ""
// icon                string        "search"         "", "search", "credit-card", "mail", "phone", "user"
// label               string        ""
// name                string        ""
// placeholder         string        ""
// required            boolean       false
// rows                string        "4"
// successMessage      string        ""
// type                string        "text"
// value               string        ""
// variant             string        "input"          "input", "textarea"

import { useState } from "react";
import FeatherIcon from "feather-icons-react/build/FeatherIcon";

const Input = ({
  arrow = true,
  disabled = false,
  errorMessage,
  hintText,
  icon = "search",
  label,
  name,
  placeholder,
  required = false,
  rows = "4",
  successMessage,
  type = "text",
  value,
  variant = "input",
}) => {
  const borderColor = errorMessage
    ? "border-error-dark"
    : "border-background-black-line";
  const fontColor = errorMessage
    ? "text-error-dark"
    : "text-type-black-primary";

  const VariantTag = variant;
  const isInput = variant === "input";
  const isTextArea = variant === "textArea";
  const iconPresent = icon && isInput;
  type = isInput ? value : "";
  value = isInput || isTextArea ? value : "";

  const [inputVal, setInputVal] = useState(value);
  const onChangeHandler = (event) => {
    setInputVal(event.target.value);
    console.log(inputVal);
  };

  return (
    <div className="flex flex-col align-start gap-[8px]">
      <div className="flex flex-row items-center justify-between"> 
        {/* Label */}
        <label
          htmlFor={name}
          className={`flex flex-row items-center p-0 text-base font-normal ${
            disabled ? "text-type-black-deactivated" : fontColor
          }`}
        >
          {label}
          <span aria-label="required">{required && "*"}</span>
        </label>

        {/* Hint Text */}
        {hintText && (
          <p
            className={`paragraph-2 ${
              disabled
                ? "text-type-black-deactivated"
                : "text-type-black-secondary"
            }`}
          >
            {hintText}
          </p>
        )}
      </div>

      <div
        className={`flex flex-row relative border focus-within:border-button-active focus-within:w-full ${borderColor} ${
          errorMessage ? "border-error-dark" : ""
        }
       ${
         disabled
           ? "bg-button-deactivatedBackground border-type-black-deactivated"
           : "text-type-black-primary"
       } `}
      >
        {/* Input / Text Area */}
        <VariantTag
          className={`peer focus:outline-none text-sm py-[14px] px-[16px] w-full ${
            iconPresent && "pl-[48px]"
          } ${isTextArea ? "resize-none" : ""}`}
          disabled={disabled}
          placeholder={placeholder}
          name={name}
          required={required}
          rows={rows}
          type={type}
          defaultValue={inputVal}
          onChange={onChangeHandler}
        ></VariantTag>

        {/* Icon */}
        {iconPresent && (
          <FeatherIcon
            icon={icon}
            size="20"
            className={`absolute pointer-events-none left-[16px] top-[14px] peer-disabled:text-type-black-deactivated text-type-black-primary`}
          />
        )}

        {/* Arrow */}
        {arrow && isInput && (
          <FeatherIcon
            icon="arrow-right"
            className={
              "pointer-events-none mr-[16px] mt-[14px] peer-disabled:text-type-black-deactivated text-type-black-primary"
            }
          />
        )}
      </div>

      {/* Error Message */}
      {errorMessage && (
        <span className="flex items-center paragraph-2 text-error-dark">
          <FeatherIcon icon="alert-triangle" size="16" className="pr-[4px]" />
          {errorMessage}
        </span>
      )}

      {/* Success Message */}
      {successMessage && (
        <span className="flex items-center paragraph-2 text-success-dark">
          <FeatherIcon icon="check-circle" size="16" className="pr-[4px]" />
          {successMessage}
        </span>
      )}
    </div>
  );
};

export default Input;

File: example/project-elixir-tailwind/lib/shipwright_elixir_tailwind_web/components/atoms/input.ex
defmodule Components.Atoms.Input do
  require Logger
  use Phoenix.Component
  attr :arrow, :boolean, default: false
  attr :disabled, :boolean, default: false
  attr :error_message, :string, default: ""
  attr :icon, :string, default: "magnifying_glass"
  attr :label, :string, default: ""
  attr :name, :string, default: ""
  attr :placeholder, :string, default: "Placeholder text"
  attr :required, :boolean, default: false
  attr :success_message, :string, default: ""
  attr :type, :string, default: "text"
  attr :hintText, :string, default: ""


  def input(assigns) do
    assigns
    |> assign_new(:borderColor, fn -> if !string_empty?(assigns[:error_message]), do: "border-error-dark", else: "border-background-black-line" end)
    |> assign_new(:fontColor, fn -> if !string_empty?(assigns[:error_message]), do: "text-error-dark", else: "text-type-black-primary" end)
    |> assign_new(:iconPresent, fn -> !string_empty?(assigns[:icon]) end)
    |> assign_new(:type, fn -> assigns[:value] end)
    |> assign_new(:value, fn -> assigns[:value] end)
    |> render
  end

  defp render(assigns) do
    ~H"""
    <div class="flex flex-col align-start gap-[8px]">
      <div class="flex flex-row items-center justify-between">
        <!-- Label -->
        <label
          for={@name}
          class={[
            "flex flex-row items-center p-0 text-base font-normal",
            @disabled && "text-type-black-deactivated"
          ]}
        >
          <%= @label %>
          <span aria-label="required">
            <%= if @required do %>
              "*"
            <% end %>
          </span>
        </label>

        <!-- Hint Text -->
        <%= if @hintText do %>
          <p
            class={[
              "paragraph-2",
              @disabled && "text-type-black-deactivated",
              !@disabled && "text-type-black-secondary"
            ]}
          >
            <%= @hintText %>
          </p>
        <% end %>
      </div>

      <div
        class={[
          "flex flex-row relative border focus-within:border-button-active focus-within:w-full",
          @borderColor,
          !string_empty?(@error_message) && "border-error-dark",
          (if @disabled, do: "bg-button-deactivatedBackground border-type-black-deactivated", else: "text-type-black-primary")
        ]}
      >
        <!-- Input / Text Area -->
        <input
          class={[
            "peer focus:outline-none text-sm py-[14px] px-[16px] w-full",
            @iconPresent && "pl-[48px]",
          ]}
          disabled={@disabled}
          placeholder={@placeholder}
          name={@name}
          required={@required}
          type={@type}
        />

        <!-- Icon -->
        <%= if @iconPresent do %>
          <Heroicons.magnifying_glass class="absolute pointer-events-none w-5 h-5 left-[16px] top-[14px] peer-disabled:text-type-black-deactivated text-type-black-primary" />
        <% end %>

        <!-- Arrow -->
        <%= if @arrow do %>
          <Heroicons.arrow_right class="absolute pointer-events-none w-5 h-5 right-[16px] top-[14px] peer-disabled:text-type-black-deactivated text-type-black-primary" />
        <% end %>
      </div>  

      <%# Error Message %>
      <%= if !string_empty?(@error_message) do %>
        <span class="flex items-center paragraph-2 text-error-dark">
          <i class="pr-[4px]" />
          <%= @error_message %>
        </span>
      <% end %>

      <%# Success Message %>
      <%= if @success_message do %>
        <span class="flex items-center paragraph-2 text-success-dark">
          <i class="pr-[4px]" />
          <%= @success_message %>
        </span>
      <% end %>
    </div>
    """
  end

  defp string_empty?(string) do
    String.length(String.trim(string)) == 0
  end
end
